// @logisticinfotech/ionic-timepicker/esm5/lib/li-ionic-timepicker.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, HostListener, ElementRef, Renderer, Input } from '@angular/core';
import { IonicTimepickerModalComponent } from './ionic-timepicker-modal/ionic-timepicker-modal.component';
import { ModalController } from '@ionic/angular';
import { NgModel, NgControl } from '@angular/forms';
var LiIonicTimepickerDirective = /** @class */ (function () {
    function LiIonicTimepickerDirective(modalCtrl, ngModel, control, el, renderer) {
        this.modalCtrl = modalCtrl;
        this.ngModel = ngModel;
        this.control = control;
        this.el = el;
        this.renderer = renderer;
        this.isModalOpen = false;
    }
    /**
     * @return {?}
     */
    LiIonicTimepickerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // console.log('this.inp', this.inputTimeConfig.clearButton);
        var _this = this;
        if (this.inputTimeConfig.clearButton !== false) {
            this.closeIcon = document.createElement('ion-icon');
            this.closeIcon.name = 'close-circle';
            this.closeIcon.className = 'clearButton';
            this.closeIcon.style.position = 'absolute';
            this.closeIcon.style.right = '8px';
            this.closeIcon.style.bottom = '30%';
            this.closeIcon.style.fontSize = '18px';
            this.closeIcon.style.zIndex = '5';
            this.closeIcon.style.color = '#A9A9A9';
            if (!this.ngModel.value || !this.control.control.value) {
                this.closeIcon.style.visibility = 'hidden';
            }
            if (this.el.nativeElement.parentNode.nodeName === 'ION-ITEM') {
                this.closeIcon.style.bottom = '12px';
            }
            this.el.nativeElement.parentNode.appendChild(this.closeIcon);
            this.renderer.listen(this.closeIcon, 'click', (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                // console.log('button clicks');
                _this.selectedTime = '';
                _this.control.control.setValue('');
                _this.ngModel.update.emit('');
            }));
        }
        /** @type {?} */
        var self = this;
        this.ngModel.valueChanges.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            // console.log('ngModel value =>', value);
            self.selectedTime = value;
            if (self.inputTimeConfig.clearButton !== false) {
                if (!value) {
                    self.closeIcon.style.visibility = 'hidden';
                }
                else {
                    self.closeIcon.style.visibility = 'visible';
                }
            }
        }));
        this.control.control.valueChanges.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            // console.log('formcontrol value =>', value);
            self.selectedTime = value;
            if (self.inputTimeConfig.clearButton !== false) {
                if (!value) {
                    self.closeIcon.style.visibility = 'hidden';
                }
                else {
                    self.closeIcon.style.visibility = 'visible';
                }
            }
        }));
        if (this.control.control.value) {
            this.selectedTime = this.control.control.value;
        }
        if (this.ngModel.value) {
            this.selectedTime = this.ngModel.value;
        }
    };
    // @HostListener('click')
    // onClick() {
    //   // console.log('on click of component =>', this.inputTimeConfig);
    //   if (!this.isModalOpen) {
    //     this.isModalOpen = true;
    //     this.openTimePicker();
    //   }
    // }
    // @HostListener('click')
    // onClick() {
    //   // console.log('on click of component =>', this.inputTimeConfig);
    //   if (!this.isModalOpen) {
    //     this.isModalOpen = true;
    //     this.openTimePicker();
    //   }
    // }
    /**
     * @return {?}
     */
    LiIonicTimepickerDirective.prototype.onFocus = 
    // @HostListener('click')
    // onClick() {
    //   // console.log('on click of component =>', this.inputTimeConfig);
    //   if (!this.isModalOpen) {
    //     this.isModalOpen = true;
    //     this.openTimePicker();
    //   }
    // }
    /**
     * @return {?}
     */
    function () {
        // console.log('on focus of component =>', this.inputTimeConfig);
        if (!this.isModalOpen) {
            this.isModalOpen = true;
            this.openTimePicker();
        }
    };
    // open time picker
    // open time picker
    /**
     * @return {?}
     */
    LiIonicTimepickerDirective.prototype.openTimePicker = 
    // open time picker
    /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var myTimePickerModal;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // console.log('open time picker modal calls');
                        return [4 /*yield*/, this.modalCtrl.create({
                                component: IonicTimepickerModalComponent,
                                cssClass: 'li-ionic-timepicker',
                                componentProps: { 'objConfig': this.inputTimeConfig, 'selectedTime': this.selectedTime }
                            })];
                    case 1:
                        myTimePickerModal = _a.sent();
                        return [4 /*yield*/, myTimePickerModal.present()];
                    case 2:
                        _a.sent();
                        myTimePickerModal.onDidDismiss()
                            .then((/**
                         * @param {?} data
                         * @return {?}
                         */
                        function (data) {
                            _this.isModalOpen = false;
                            // console.log(data);
                            if (data.data && data.data.time) {
                                _this.selectedTime = data.data.time;
                                _this.control.control.setValue(_this.selectedTime);
                                _this.ngModel.update.emit(_this.selectedTime);
                            }
                        }));
                        return [2 /*return*/];
                }
            });
        });
    };
    LiIonicTimepickerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[liIonicTimepicker]',
                    exportAs: 'liIonicTimepicker',
                    providers: [NgModel]
                },] }
    ];
    /** @nocollapse */
    LiIonicTimepickerDirective.ctorParameters = function () { return [
        { type: ModalController },
        { type: NgModel },
        { type: NgControl },
        { type: ElementRef },
        { type: Renderer }
    ]; };
    LiIonicTimepickerDirective.propDecorators = {
        inputTimeConfig: [{ type: Input, args: ['liIonicTimepicker',] }],
        onFocus: [{ type: HostListener, args: ['ionFocus',] }]
    };
    return LiIonicTimepickerDirective;
}());
export { LiIonicTimepickerDirective };
if (false) {
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.inputTimeConfig;
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.isModalOpen;
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.closeIcon;
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.selectedTime;
    /**
     * @type {?}
     * @private
     */
    LiIonicTimepickerDirective.prototype.modalCtrl;
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.ngModel;
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.control;
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.el;
    /** @type {?} */
    LiIonicTimepickerDirective.prototype.renderer;
}
